src/main/java/demo/ReadBalancingDataSource.java

```diff
@@ -32,18 +32,25 @@ public ReadDataSource(DataSource dataSource) {
         DataSource dataSource;
         AtomicLong blacklistedUntil;
 
-        public boolean isBlackListed() {
-            long blacklistedUntilTime = this.blacklistedUntil.get();
-            
-            if (blacklistedUntilTime == -1) {
-                return false;
-            }
-            
-            if (blacklistedUntilTime > System.currentTimeMillis()) {
-                this.blacklistedUntil.set(-1);
-                return true;
-            }
-            return false ;
+        public boolean isBlacklisted() {
+            long currentBlacklistedUntil;
+            do {
+                currentBlacklistedUntil = this.blacklistedUntil.get();
+
+                if (currentBlacklistedUntil == -1) {
+                    return false;
+                }
+
+                if (currentBlacklistedUntil < blacklistClockTime()) {
```

Comment:
it checked if we are still black listed so I believe it's right no?
currentBlacklistedUntil > currentTime then we are passed it ?